frontend/src/app/user/register/register.component.ts

```diff
@@ -0,0 +1,100 @@
+import { Component, DestroyRef, OnInit } from '@angular/core';
+import { UserRegistration } from '../models/user';
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+import { usernameValidators, passwordValidators, identicalPasswordsValidator } from '../validation';
+import { AuthService } from '@services/auth.service';
+import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
+import { map, merge, startWith } from 'rxjs';
+import { controlErrorMessages$, formErrorMessages$, setErrors, updateFormValidity } from '../utils';
+import { ToastService } from '@services/toast.service';
+import { Router } from '@angular/router';
+
+type RegisterForm = {
+    [key in keyof UserRegistration]: FormControl<UserRegistration[key]>;
+}
+
+@Component({
+    selector: 'lc-register',
+    templateUrl: './register.component.html',
+    styleUrls: ['./register.component.scss']
+})
+export class RegisterComponent implements OnInit {
+    public form = new FormGroup<RegisterForm>({
+        username: new FormControl<string>('', {
+            nonNullable: true,
+            validators: [
+                Validators.required,
+                ...usernameValidators
+            ]
+        }),
+        email: new FormControl<string>('', {
+            nonNullable: true,
+            validators: [
+                Validators.required,
+                Validators.email,
+            ]
+        }),
+        password1: new FormControl<string>('', {
+            nonNullable: true,
+            validators: [
+                Validators.required,
+                ...passwordValidators,
+            ]
+        }),
+        password2: new FormControl<string>('', {
+            nonNullable: true,
+            validators: [
+                Validators.required,
+                ...passwordValidators,
+            ]
+        }),
+    }, {
+        validators: identicalPasswordsValidator<keyof RegisterForm>('password1', 'password2')
+    });
+
+    public usernameErrors$ = controlErrorMessages$(this.form, 'username');
```

Comment:
Instead of creating this stream of error messages here, we could also create a pipe that extracts them from a control in the template.

That could simplify this TS file, but again, I did not want to unnecessarily introduce another layer of complexity where it's not strictly needed.